### PR TITLE
first attempt at making the MWU test not stochastic

### PR DESCRIPTION
**Pull request by jpaul6**

### CODE REVIEW

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java [hunk 1]
```diff
@@ -82,7 +78,18 @@ public MannWhitneyU(ExactMode mode) {
...
+            observations.put(n, currValue);
+        }
+
+        if (set == USet.SET1) {
+            currValue.first += 1;
+        } else {
+            currValue.second += 1;
+        }
+
         if ( set == USet.SET1 ) {
```
Comment (caethan) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java]: These checks can be done simultaneously

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java [hunk 2]
```diff
@@ -154,18 +163,18 @@ public void add(Number n, USet set) {
      */
```
Comment (caethan) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java:159]: Should update the comments here for what tieStructure is supposed to be

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java [hunk 3]
```diff
@@ -154,18 +163,18 @@ public void add(Number n, USet set) {
...
     @Requires({"m > 0","n > 0"})
     @Ensures({"result != null", "! Double.isInfinite(result.getFirst())", "! Double.isInfinite(result.getSecond())"})
-    protected static Pair<Double,Double> calculateP(int n, int m, long u, boolean twoSided, ExactMode exactMode) {
+    protected static Pair<Double,Double> calculateP(int n, int m, double u, boolean twoSided, Map<Integer, Integer> tieStructure, ExactMode exactMode) {
         Pair<Double,Double> zandP;
         if ( n > 8 && m > 8 ) {
             // large m and n - normal approx
-            zandP = calculatePNormalApproximation(n,m,u, twoSided);
+            zandP = calculatePNormalApproximation(n,m,u,tieStructure, twoSided);
         } else if ( n > 5 && m > 7 ) {
```
Comment (caethan) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java:167]: Huh.  This is...weird.  The bounding regions are weird and it's not actually using the uniform approximation.  Why not just clean up the if cases and add a TODO comment in that the uniform approximation is worth looking into?

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java [hunk 4]
```diff
@@ -208,9 +217,16 @@ public void add(Number n, USet set) {
      */
     @Requires({"m > 0","n > 0"})
     @Ensures({"! Double.isNaN(result)", "! Double.isInfinite(result)"})
-    private static double getZApprox(int n, int m, long u) {
+    private static double getZApprox(int n, int m, double u, Map<Integer, Integer> tieStructure) {
         double mean = ( ((long)m)*n+1.0)/2;
```
Comment (caethan) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java]: Is there some off-by-one issue here?  The mean value of the normal distribution should be nm/2, not (nm + 1) / 2.

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java [hunk 5]
```diff
@@ -208,9 +217,16 @@ public void add(Number n, USet set) {
...
         double var = (((long) n)*m*(n+m+1.0))/12;
+        if (tieStructure != null) {
+            double varAdjust = 0;
+            for (Map.Entry<Integer, Integer> tie : tieStructure.entrySet()) {
+                varAdjust += tie.getValue() * (Math.pow(tie.getKey(), 3) - tie.getKey());
+            }
+            var -= varAdjust / ((n + m) * (n + m - 1));
```
Comment (caethan) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java]: Your adjustment is off here - you need

var -= varAdjust \* (n \* m) / (12 \* (n + m) \* (n + m - 1))

(It's an adjustment to the n + m + 1 term in the initial calculation)

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java (hunk 2 above)
Comment (jpaulLD) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java:159]: I'll add docs for the generating method.

File: public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java (hunk 3 above)
Comment (jpaulLD) [inline public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MannWhitneyU.java:167]: I'm going to leave this so as not to muddy up the diff when we PR to broad.

**Pull request merged**